File: packages/gbnf-experiment/data/20260909T180307Z_280f3890/ported_implementation/src/utils/errors/build-error-position.ts
export const MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW = 3;

export const buildErrorPosition = (src: string, pos: number): string[] => {
  if (src === '') {
    return ['No input provided'];
  }
  const lines = src.split('\n');

  let lineIdx = 0;
  let line = lines[lineIdx];
  while (line && pos > line.length - 1 && pos < src.length) {
    pos -= line.length;
    lineIdx += 1;
    line = lines[lineIdx];
  }

  const linesToShow = lines.slice(
    Math.max(0, lineIdx - (MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW - 1)),
    lineIdx + 1,
  );

  return [...linesToShow, `${' '.repeat(Math.max(0, pos))}^`];
};

File: packages/gbnf-experiment/data/20260909T173101Z_8e3da990/ported_implementation/utils/errors/build_error_position.py
MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW = 3


def build_error_position(src: str, pos: int) -> list[str]:
    if src == "":
        return ["No input provided"]

    lines = src.split('\n')

    line_idx = 0
    while line_idx < len(lines) and lines[line_idx] and pos > len(lines[line_idx]) - 1:
        pos -= len(lines[line_idx])
        line_idx += 1

    lines_to_show: list[str] = []
    for i in range(max(0, line_idx - (MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW - 1)), line_idx + 1):
        # past the end of the grammar the reference pushes `undefined`, which
        # joins as an empty line.
        lines_to_show.append(lines[i] if i < len(lines) else '')
    return [
        *lines_to_show,
        ' ' * pos + '^',
    ]


buildErrorPosition = build_error_position
